core/src/apps/cardano/address.py

```diff
@@ -1,137 +1,249 @@
-    The max value for `a` is 20, 1 000 000 for `i`.
-    The derivation scheme v1 allowed a'/0/i only,
-    but in v2 it can be a'/1/i as well.
-    """
-    if len(path) != 5:
-        return False
-    if path[0] != 44 | HARDENED:
-        return False
-    if path[1] != 1815 | HARDENED:
-        return False
-    if path[2] < HARDENED or path[2] > 20 | HARDENED:
-        return False
-    if path[3] != 0 and path[3] != 1:
-        return False
-    if path[4] > 1000000:
-        return False
-    return True
 
+def _create_address_header(address_type: int, network_id: int) -> bytes:
+    if address_type == CardanoAddressType.BYRON_ADDRESS:
+        """
+        Byron addresses don't have an explicit header in the Shelley
+        spec. However, thanks to their CBOR structure they always start with
+        0b1000 - the byron address id. This is no coincidence.
+        The Shelley address headers are purposefully built around these
+        starting bits of the byron address.
+        """
+        raise ValueError("Byron address does not contain an explicit header!")
+
+    address_id = _get_address_id(address_type)
+    header = address_id << 4 | network_id
 
-def _address_hash(data) -> bytes:
-    data = cbor.encode(data)
-    data = hashlib.sha3_256(data).digest()
-    res = hashlib.blake2b(data=data, outlen=28).digest()
-    return res
+    return bytes([header])
 
 
-def _get_address_root(node, payload):
-    extpubkey = remove_ed25519_prefix(node.public_key()) + node.chain_code()
-    if payload:
-        payload = {1: cbor.encode(payload)}
+def _get_address_id(address_type: int) -> int:
+    if address_type == CardanoAddressType.BYRON_ADDRESS:
+        address_id = AddressId.BYRON_ADDRESS
+    elif address_type == CardanoAddressType.BASE_ADDRESS:
+        address_id = AddressId.BASE_ADDRESS
+    elif address_type == CardanoAddressType.POINTER_ADDRESS:
+        address_id = AddressId.POINTER_ADDRESS
+    elif address_type == CardanoAddressType.ENTERPRISE_ADDRESS:
+        address_id = AddressId.ENTERPRISE_ADDRESS
+    elif address_type == CardanoAddressType.REWARD_ADDRESS:
+        address_id = AddressId.REWARD_ADDRESS
     else:
-        payload = {}
-    return _address_hash([0, [0, extpubkey], payload])
+        raise ValueError("Invalid address type '%s'!" % address_type)
+
+    return address_id
+
+
+def _derive_base_address(
+    keychain: seed.Keychain,
+    address_header: bytes,
+    path: List[int],
+    staking_key_path: List[int],
+    staking_key_hash: bytes,
+) -> str:
+    spending_key_hash = get_public_key_hash(keychain, path)
+
+    _validate_base_address_staking_info(staking_key_path, staking_key_hash)
+
+    if staking_key_hash is None:
+        staking_key_hash = get_public_key_hash(keychain, staking_key_path)
+
+    return address_header + spending_key_hash + staking_key_hash
+
+
+def _validate_base_address_staking_info(
+    staking_key_path: List[int], staking_key_hash: bytes,
+):
+    if staking_key_hash is None:
```

Comment:
if nesting is bad. prefer early-exit / dealin with edge cases first approach

Reply:
Edited.